File: sdk/ios/Sources/MosaicSDK/MockPurchaseProvider.swift
import Foundation

public enum MockMosaicPurchaseBehavior: Sendable, Equatable {
  /// Succeeds once, then reports the product as already entitled.
  case automatic
  case success
  case cancellation
  case failure(diagnosticCode: String)
  case unavailable
  case alreadyEntitled
}

public enum MockMosaicRestoreBehavior: Sendable, Equatable {
  /// Returns current mock entitlements, or `nothingToRestore` when empty.
  case automatic
  case success(Set<MosaicEntitlement>)
  case alreadyEntitled(Set<MosaicEntitlement>)
  case noPurchases
  case failure(diagnosticCode: String)
}

/// Deterministic provider for Phase 1 examples and tests; it never talks to a
/// store and is not a StoreKit simulator or production adapter.
public actor MockMosaicPurchaseProvider: MosaicPurchaseProvider {
  private let products: [String: MosaicProduct]
  private let purchaseBehavior: MockMosaicPurchaseBehavior
  private let restoreBehavior: MockMosaicRestoreBehavior
  private var entitlements: Set<MosaicEntitlement>

  public init(
    products: [MosaicProduct] = [],
    activeEntitlements: Set<MosaicEntitlement> = [],
    purchaseBehavior: MockMosaicPurchaseBehavior = .automatic,
    restoreBehavior: MockMosaicRestoreBehavior = .automatic
  ) {
    self.products = Dictionary(
      products.map { ($0.id, $0) },
      uniquingKeysWith: { _, replacement in replacement }
    )
    self.purchaseBehavior = purchaseBehavior
    self.restoreBehavior = restoreBehavior
    entitlements = activeEntitlements
  }

  public func loadProducts(identifiers: [String]) async -> MosaicProductLoadResult {
    let available = identifiers.compactMap { products[$0] }
    if available.isEmpty, !identifiers.isEmpty {
      return .unavailable(
        productIDs: identifiers,
        diagnosticCode: "mock_products_unavailable"
      )
    }
    return .loaded(available)
  }

  public func purchase(productID: String) async -> MosaicPurchaseResult {
    guard products[productID] != nil else {
      return .productUnavailable(productID: productID)
    }

    switch purchaseBehavior {
    case .automatic:
      let entitlement = MosaicEntitlement(id: productID)
      guard !entitlements.contains(entitlement) else {
        return .alreadyEntitled(productID: productID)
      }
      entitlements.insert(entitlement)
      return .purchased(productID: productID, transactionID: "mock-\(productID)")
    case .success:
      entitlements.insert(MosaicEntitlement(id: productID))
      return .purchased(productID: productID, transactionID: "mock-\(productID)")
    case .cancellation:
      return .cancelled(productID: productID)
    case .failure(let diagnosticCode):
      return .failed(productID: productID, diagnosticCode: diagnosticCode)
    case .unavailable:
      return .productUnavailable(productID: productID)
    case .alreadyEntitled:
      entitlements.insert(MosaicEntitlement(id: productID))
      return .alreadyEntitled(productID: productID)
    }
  }

  public func restore() async -> MosaicRestoreResult {
    switch restoreBehavior {
    case .automatic:
      return entitlements.isEmpty ? .nothingToRestore : .restored(entitlements)
    case .success(let restored):
      entitlements.formUnion(restored)
      return .restored(restored)
    case .alreadyEntitled(let existing):
      entitlements.formUnion(existing)
      return .alreadyEntitled(existing)
    case .noPurchases:
      return .nothingToRestore
    case .failure(let diagnosticCode):
      return .failed(diagnosticCode: diagnosticCode)
    }
  }

  public func activeEntitlements() async -> MosaicActiveEntitlementsResult {
    .active(entitlements)
  }
}

extension MosaicProduct {
  public static let phase1Monthly = MosaicProduct(
    id: "mosaic_pro_monthly",
    title: "Mosaic Pro Monthly",
    localizedPrice: "$7.99",
    localizedSubscriptionPeriod: "per month"
  )

  public static let phase1Yearly = MosaicProduct(
    id: "mosaic_pro_yearly",
    title: "Mosaic Pro Yearly",
    localizedPrice: "$49.99",
    localizedSubscriptionPeriod: "per year"
  )

  public static let phase1MockProducts: [MosaicProduct] = [
    .phase1Monthly,
    .phase1Yearly,
  ]
}

File: sdk/android/mosaic/src/main/kotlin/dev/mosaic/sdk/MockPurchaseProvider.kt
package dev.mosaic.sdk

enum class MosaicMockPurchaseScenario {
    SUCCESS,
    CANCELLED,
    FAILURE,
    PRODUCT_UNAVAILABLE,
    ALREADY_ENTITLED,
}

enum class MosaicMockRestoreScenario {
    AUTOMATIC,
    RESTORED,
    ALREADY_ENTITLED,
    NOTHING_TO_RESTORE,
    FAILURE,
}

/** Deterministic provider for examples and tests; it never talks to a store. */
class MockMosaicPurchaseProvider(
    products: List<MosaicProduct> = emptyList(),
    activeEntitlements: Set<MosaicEntitlement> = emptySet(),
    private val purchaseScenario: MosaicMockPurchaseScenario = MosaicMockPurchaseScenario.SUCCESS,
    private val restoreScenario: MosaicMockRestoreScenario = MosaicMockRestoreScenario.AUTOMATIC,
    private val restoredEntitlements: Set<MosaicEntitlement> = setOf(MosaicEntitlement("mosaic-pro")),
) : MosaicPurchaseProvider {
    private val lock = Any()
    private val productsById = products.associateBy(MosaicProduct::id)
    private val entitlements = activeEntitlements.toMutableSet()

    override suspend fun loadProducts(productIds: List<String>): MosaicProductLoadResult =
        synchronized(lock) {
            val loaded = productIds.mapNotNull(productsById::get)
            val missing = productIds.filterNot(productsById::containsKey)
            if (loaded.isEmpty() && missing.isNotEmpty()) {
                MosaicProductLoadResult.Unavailable(productIds = missing)
            } else {
                MosaicProductLoadResult.Loaded(
                    products = loaded,
                    unavailableProductIds = missing.toSet(),
                )
            }
        }

    override suspend fun purchase(productId: String): MosaicPurchaseResult = synchronized(lock) {
        if (!productsById.containsKey(productId)) {
            return@synchronized MosaicPurchaseResult.ProductUnavailable(productId)
        }
        when (purchaseScenario) {
            MosaicMockPurchaseScenario.SUCCESS -> {
                val entitlement = MosaicEntitlement(productId)
                if (!entitlements.add(entitlement)) {
                    MosaicPurchaseResult.AlreadyEntitled(productId)
                } else {
                    MosaicPurchaseResult.Purchased(
                        productId = productId,
                        transactionId = "mock-$productId",
                    )
                }
            }
            MosaicMockPurchaseScenario.CANCELLED -> MosaicPurchaseResult.Cancelled(productId)
            MosaicMockPurchaseScenario.FAILURE -> MosaicPurchaseResult.Failed(productId)
            MosaicMockPurchaseScenario.PRODUCT_UNAVAILABLE ->
                MosaicPurchaseResult.ProductUnavailable(productId)
            MosaicMockPurchaseScenario.ALREADY_ENTITLED ->
                MosaicPurchaseResult.AlreadyEntitled(productId)
        }
    }

    override suspend fun restore(): MosaicRestoreResult = synchronized(lock) {
        when (restoreScenario) {
            MosaicMockRestoreScenario.AUTOMATIC -> {
                if (entitlements.isEmpty()) {
                    MosaicRestoreResult.NothingToRestore
                } else {
                    MosaicRestoreResult.Restored(entitlements.toSet())
                }
            }
            MosaicMockRestoreScenario.RESTORED -> {
                entitlements += restoredEntitlements
                MosaicRestoreResult.Restored(restoredEntitlements)
            }
            MosaicMockRestoreScenario.ALREADY_ENTITLED -> {
                entitlements += restoredEntitlements
                MosaicRestoreResult.AlreadyEntitled(restoredEntitlements)
            }
            MosaicMockRestoreScenario.NOTHING_TO_RESTORE -> MosaicRestoreResult.NothingToRestore
            MosaicMockRestoreScenario.FAILURE -> MosaicRestoreResult.Failed()
        }
    }

    override suspend fun activeEntitlements(): MosaicActiveEntitlementsResult = synchronized(lock) {
        MosaicActiveEntitlementsResult.Active(entitlements.toSet())
    }

    companion object {
        fun phase1Products(): List<MosaicProduct> = listOf(
            MosaicProduct(
                id = "mosaic_pro_monthly",
                title = "Mosaic Pro Monthly",
                localizedPrice = "$9.99",
                subscriptionPeriod = "month",
            ),
            MosaicProduct(
                id = "mosaic_pro_yearly",
                title = "Mosaic Pro Yearly",
                localizedPrice = "$79.99",
                subscriptionPeriod = "year",
            ),
            MosaicProduct(
                id = "mosaic_pro_lifetime",
                title = "Mosaic Pro Lifetime",
                localizedPrice = "$199.99",
            ),
        )
    }
}
